jetty-servlet/src/test/java/org/eclipse/jetty/servlet/AsyncServletIOTest.java

```diff
@@ -787,5 +798,187 @@ public void onAllDataRead() throws IOException
         }
     }
     
+
+    @Test
+    public void testStolenAsyncRead() throws Exception
+    {
+        StringBuilder request = new StringBuilder(512);
+        request.append("POST /ctx/stolen/info HTTP/1.1\r\n")
+        .append("Host: localhost\r\n")
+        .append("Content-Type: text/plain\r\n")
+        .append("Content-Length: 2\r\n")
+        .append("\r\n")
+        .append("1");
+        int port=_port;
+        List<String> list = new ArrayList<>();
+        try (Socket socket = new Socket("localhost",port))
+        {
+            socket.setSoTimeout(10000);
+            OutputStream out = socket.getOutputStream();
+            out.write(request.toString().getBytes(ISO_8859_1));
+            out.flush();
+            
+            // wait until server is ready
+            _servletStolenAsyncRead.ready.await();
+            final CountDownLatch wait = new CountDownLatch(1);
+            
+            // Stop any dispatches until we want them
+            Function<Runnable,Runnable> old = _wQTP.wrapper.getAndSet(r->
+            { 
+                return new Runnable()
+                {
+                    public void run() {
+                        try
+                        {
+                            wait.await();
+                            r.run();
+                        }
+                        catch (InterruptedException e)
+                        {
+                            e.printStackTrace();
+                        }
+                    }
+                };
+            });
+
+            // We are an unrelated thread, let's mess with the input stream
+            ServletInputStream sin = _servletStolenAsyncRead.listener.in;
+            sin.setReadListener(_servletStolenAsyncRead.listener);
+            // thread should be dispatched to handle, but held by our wQTP wait.
+            
+            // Let's steal our read
+            Assert.assertTrue(sin.isReady());
+            Assert.assertThat(sin.read(),Matchers.is((int)'1'));
+            Assert.assertFalse(sin.isReady());
+            
+            // let the ODA call go
+            wait.countDown();
+            _wQTP.wrapper.set(old);
```

Comment:
I think these 2 lines should be switched: you don't want to count down the latch which may produce something for the `wQTP` which would still have the replaced wrapper - while you want it to have the original one.